meta-facebook/yv35-rf/src/platform/plat_mctp.c

```diff
@@ -0,0 +1,214 @@
+ * You may obtain a copy of the License at
+ * 
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Copyright (c) 2012-2014 Wind River Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <sys/printk.h>
+#include <logging/log.h>
+#include <logging/log_ctrl.h>
+#include <stdlib.h>
+#include "mctp.h"
+#include "mctp_ctrl.h"
+#include "pldm.h"
+#include "ipmi.h"
+#include "sensor.h"
+#include "plat_hook.h"
+#include "plat_mctp.h"
+#include "plat_gpio.h"
+#include "cci.h"
+#include "pm8702.h"
+
+LOG_MODULE_REGISTER(plat_mctp);
+K_TIMER_DEFINE(send_cmd_timer, send_cmd_to_dev, NULL);
+K_WORK_DEFINE(send_cmd_work, send_cmd_to_dev_handler);
+
+static mctp_smbus_port smbus_port[] = {
+	{ .conf.smbus_conf.addr = I2C_ADDR_BIC, .conf.smbus_conf.bus = I2C_BUS_CXL },
+};
+
+static mctp_route_entry mctp_route_tbl[] = {
+	{ CXL_EID, I2C_BUS_CXL, I2C_ADDR_CXL0 },
+};
+
+mctp *find_mctp_by_smbus(uint8_t bus)
+{
+	uint8_t i;
+	for (i = 0; i < ARRAY_SIZE(smbus_port); i++) {
+		mctp_smbus_port *p = smbus_port + i;
+		if (bus == p->conf.smbus_conf.bus) {
+			return p->mctp_inst;
+		}
+	}
+	return NULL;
+}
+
+static void set_endpoint_resp_handler(void *args, uint8_t *buf, uint16_t len)
+{
+	if (!buf || !len)
+		return;
+	LOG_HEXDUMP_INF(buf, len, __func__);
+}
+
+static void set_endpoint_resp_timeout(void *args)
+{
+	mctp_route_entry *p = (mctp_route_entry *)args;
+	printk("[%s] Endpoint 0x%x set endpoint failed on bus %d \n", __func__, p->endpoint,
+	       p->bus);
+}
+
+static void set_dev_endpoint(void)
+{
+	for (uint8_t i = 0; i < ARRAY_SIZE(mctp_route_tbl); i++) {
+		mctp_route_entry *p = mctp_route_tbl + i;
+		for (uint8_t j = 0; j < ARRAY_SIZE(smbus_port); j++) {
+			if (p->bus != smbus_port[j].conf.smbus_conf.bus) {
+				continue;
+			}
+			printk("Prepare send endpoint bus 0x%x, addr 0x%x\n", p->bus, p->addr);
+			struct _set_eid_req req = { 0 };
+			req.op = SET_EID_REQ_OP_SET_EID;
+			req.eid = p->endpoint;
+
+			mctp_ctrl_msg msg;
+			memset(&msg, 0, sizeof(msg));
+			msg.ext_params.type = MCTP_MEDIUM_TYPE_SMBUS;
+			msg.ext_params.smbus_ext_params.addr = p->addr;
+
+			msg.hdr.cmd = MCTP_CTRL_CMD_SET_ENDPOINT_ID;
+			msg.hdr.rq = 1;
+
+			msg.cmd_data = (uint8_t *)&req;
+			msg.cmd_data_len = sizeof(req);
+
+			msg.recv_resp_cb_fn = set_endpoint_resp_handler;
+			msg.timeout_cb_fn = set_endpoint_resp_timeout;
+			msg.timeout_cb_fn_args = p;
+
+			mctp_ctrl_send_msg(find_mctp_by_smbus(p->bus), &msg);
+		}
+	}
+}
+
+static uint8_t mctp_msg_recv(void *mctp_p, uint8_t *buf, uint32_t len, mctp_ext_params ext_params)
+{
+	if (!mctp_p || !buf || !len)
```

Comment:
Please use CHECK_NULL_ARG_WITH_RETURN to test for NULL values.

Also, is len always expected to be zero or is it supposed to be a pointer to len (e.g. uint32_t *len)?